### PR TITLE
[RFC] nginx-util: use uci for server configuraton

### PR DESCRIPTION
**Pull request by peter-stadler**
Maintainer: me
Compile tested: x86_64, x86_64 qemu, master
Run tested: x86_64, x86_64 qemu, master, running test-nginx-ssl-util-root.sh

**tl;dr:** The functions `{add,del}_ssl` modify a server section of the UCI config if there is no `.conf` file with the same name in `/etc/nginx/conf.d/`.

Then `init_lan` creates `/var/lib/nginx/uci.conf` files by copying the `/etc/nginx/uci.conf.template` and standard options from the UCI config; additionally the special path `logd` can be used in `{access,error}_log`.

The init does not change the configuration beside re-creating self-signed certificates when needed. This is also the only purpose of the new `check_ssl`, which is installed as yearly cron job.

**Initialization:**

Invoking `nginx-util init_lan` parses the UCI configuration for package `nginx`. It creates a server part in `/var/lib/nginx/uci.conf` for every `section server '$name'` by copying all UCI options but the following:

* `option uci_manage_ssl` is skipped. It is set to 'self-signed' by `nginx-util add_ssl $name`, removed by `nginx-util del_ssl $name` and used by `nginx-util check_ssl` (see below).

* `logd` as path in `error_log` or `access_log` writes them to STDERR respective STDOUT, which are fowarded by Nginx's init to the log daemon. Specifically:
`option error_log 'logd'` becomes `error_log stderr;` and
`option access_log 'logd openwrt'` becomes `access_log /proc/self/fd/1 openwrt;`

* Other `[option|list] key 'value'` entries just become `key value;` directives.

The init calls internally also `check_ssl` for rebuilding self-signed SSL certificates if needed (see below). And it still sets up `/var/lib/nginx/lan{,_ssl}.listen` files as it is doing in the current version (so they stay available).

**Defaults:**

The package installs the file `/etc/nginx/restrict_locally` containing allow/deny directives for restricting the access to LAN addresses by including it into a server part. This file is included by the default server that listens on all IPs (instead of only the local IPs as it did before, other servers do not need to listen on the local IPs explicitly anymore). The default server is contained (together with a server that redirects inexistent URLs to https in the SSL version) in the UCI configuration file `/etc/config/nginx`. Furthermore, the packages installs a `/etc/nginx/uci.conf.template` containing the current setup and a marker, which will be replaced by the created UCI servers when calling `init_lan`.

**Other:**

If there is a file named `/etc/nginx/conf.d/$name.conf` the functions `init_lan`, `add_ssl $name` and `del_ssl $name` will use that file instead of a UCI server section with the same $name (as it is doing in the current version).

Else if there is only a UCI `section server $name`:
* `nginx-util add_ssl $name` will add to it:
`option uci_manage_ssl 'self-signed'`
`option ssl_certificate '/etc/nginx/conf.d/$name.crt'`
`option ssl_certificate_key '/etc/nginx/conf.d/$name.key'`
`option ssl_session_cache 'shared:SSL:32k'`
`option ssl_session_timeout '64m'`
If these options are already present, they will stay the same; just the first option `uci_manage_ssl` will always be changed to 'self-signed'. The command also changes all `listen` list items to use port 443 and ssl instead of port 80 (without ssl). If they stated another port than 80 before, they are kept the same. Furthermore, it creates a self-signed SSL certificate if necessary, i.e., if there is no *valid* certificate and key at the locations given by the options `ssl_certificate` and `ssl_certificate_key`.

* `nginx-util del_ssl $name` checks if `uci_manage_ssl` is set 'self-signed' in the corresponding UCI section. Only then it removes all of the above options regardless of the value looking just at the key name. Then, it also changes all `listen` list items to use port 80 (without ssl) instead of port 443 with ssl. If stating another port than 443, they are kept the same. Furthermore, it removes the SSL certificate and key that were indicated by `ssl_certificate{,_key}`.

* `nginx-util check_ssl` looks through all server sections of the UCI config for `uci_manage_ssl 'self-signed'`. On every hit it checks if the SSL certificate-key-pair indicated by the options `ssl_certificate{,_key}` is expired. Then it re-creates a self-signed certificate. If there exists at least one `section server` with `uci_manage_ssl 'self-signed'`, it will try to install itself as cron job. If there are no such sections, it removes that cron job if possible.

### Plan

This change prepares Nginx for using UCI. So, the decisions here will influence Nginx's behavior. 

**Configuration:**

I (still) think it is the best to give a standard configuration with a default server that uses a self-signed certificate (if SSL is enabled) <strike>and listens locally</strike> but listens on all IPs restricting the access locally. <strike>(This implies that other servers must also listen explicitly on local IPs to be reachable from LAN.)</strike>

I do not see a possibility to achieve this without adding some configuration, which could interfere with existing custom configuration. To make it more smoothly, it would not use the new `_lan` default server from the `/etc/config/nginx` UCI configuration if the file `/etc/nginx/conf.d/_lan.conf` is still there (it will only be removed in a future change if it was not modified).

This is done in the UCI configuration `/etc/config/nginx`. There are two versions of this file depending on whether Nginx is installed with or without SSL support, which could also be installed by the nginx package.

Although, we could make the default server optional package(s), this would be different to all previous versions (there was always a default server, although there were differences if LuCI using Nginx was installed or not). Then other packages (e.g. `luci-nginx`, `luci-ssl-nginx`, `ariang` or `etesync-server`) could depend on the new package(s), and the default server is only installed then.

IMHO we should *not* separate the default server in other package(s): This would only postpone problems of interference with previous configurations to the point when such a package is installed (and then it becomes more difficult to spot). And since this would be a bigger change, I would only do it if a maintainer of Nginx (@Ansuel and @heil) want it.

**Main Functions:**

The main purpose of this tool is to unify the workload of `/etc/init.d/nginx` with the functionality of `px5g` by the commands:
* `init_lan` sets up the configuration for Nginx (faster than doing all in a shell script `/etc/init.d/nginx` that calls this function)
* `add_ssl name' creates a self-signed certificates (like `px5g` but linked against `libopenssl`) setting up corresponding SSL directives in Nginx's config, too.
* `del_ssl name` undoes that.

Additionally there is a yearly cron job (if needed and cron is available) using the function:
* `check_ssl` re-creates self-signed certificates if they would expire (in the next 13 months).

This is how to interact with server $name:
* delete it from config: `uci delete nginx.$name`
* add self-signed SSL: `nginx-util add_ssl $name`
* disable re-creation of self-signed certificate: `uci delete nginx.$name.uci_manage_ssl` or better:
* remove self-signed SSL: `nginx-util del_ssl $name`
* reload (all) config: `service nginx reload`

**Proposed Timeline:**

1. Discuss these changes (until 11th of March draft PR)
2. Change and test these tools (until 22nd of March make PR)
3. Adopt the [wiki documentation](https://openwrt.org/docs/guide-user/services/webserver/nginx#configuration) (until end of March)
4. Change Nginx to setup the default server via UCI (using `/etc/nginx/uci.conf` if it is a symlink to `/var/lib/nginx/uci.conf` and removing `/etc/nginx/conf.d/_lan.conf` and `/etc/nginx/conf.d/_redirect2ssl.conf` for the SSL version until 1st of April make PR)

Maybe, I will look into making some Nginx options available through LuCI afterwards.

### Questions

There are some things, I am not sure how to do:
* Should it use other directory/file names, e.g. `/var/etc/nginx/...` for dynamic configurations?
* <strike>Should we install the UCI config with this package or with Nginx?</strike> (Here) <strike> Should we use `/etc/config/nginx-util` instead of `/etc/config/nginx` in the first case?</strike> (No)
* <strike>How should we handle a `stream { … }` part in the Nginx configuration? Currently you have to edit the main `/etc/nginx/nginx.conf` file, which will not get updated if it was modified.</strike> (Setup manually.)

### To be changed (updated according to discussion)

**Dynamically create the main configuration file and symlink `/etc/nginx/uci.conf` to it.**

Prepared and edited above description: I changed it to include a `/etc/nginx/uci.conf.template` and to create `/var/lib/nginx/uci.conf` from it via `nginx-util init_lan` using servers of the UCI config in `/etc/config/nginx`.

I would suggest to install the symlink by the Nginx package when adopting it to the new version of nginx-util.

**In the testscripts use /bin/sh and shellcheck them.**

Done.

**Use allow/deny instead of listen locally**

Done and edited the above description: there will be no `uci_listen_locally` …

### CODE REVIEW

File: net/nginx-util/src/nginx-util.cpp [hunk 1]
```diff
@@ -1,103 +1,185 @@
...
+
+
+void init_conf(const uci::package & pkg, const std::vector<std::string> & ips)
+{
+    for (auto sec : pkg) {
+        if (sec.type()!=std::string_view{"server"}) { continue; } //else:
+
+        auto secname = sec.name();
+
+        auto legacypath = std::string{CONF_DIR} + secname + ".conf";
```
Comment (neheb) [inline net/nginx-util/src/nginx-util.cpp]: No idea if this would be overengineering, but would std::filesystem be useful here?

Comment (peter-stadler) [reply net/nginx-util/src/nginx-util.cpp]: I appreciate all your suggestions a lot, but I am not sure if I should use it here: It would be useful, but it increases the .ipk size by about 40kB, e.g. `nginx-ssl-util` from 58kB to 98kB.
Do you think the benefits outweigh it? Then I would change it (but maybe in a separate PR afterwards as it would affect different places in the code).

**Comment by Ansuel** [comment]
I think that if the uci support is used for nginx, then we should follow the wa we already use for samba,unbound and others...
Add an option to use the normal nginx.conf and use separate dir for the dynamic nginx configuration 

I think the uci config file should installed by this package instead of the normal nginx package

Why should we change the stream part?

**Comment by peter-stadler** [comment]
Right now the UCI config covers only server parts, so we will use always the `/etc/nginx/nginx.conf` main configuration.

There will be a new file `_uci.conf` in `/etc/nginx/conf.d/` that is include by the main `nginx.conf` and includes all dynamic files.

The dynamic files should live on the tmpfs `/var/`, the question is if `/var/lib/nginx/*.conf` is the right place or if we should use e.g. `/etc/nginx/*.conf` (despite `/var/lib/nginx/` will still be there)?

Should we add an option to the UCI config to ignore other `/etc/nginx/conf.d/*.conf` files?

Some users use the stream part, should we change the main config adding a `include /etc/nginx/stream.d/*.conf` for example or should we leave it to the users?

**Comment by Ansuel** [comment]
since it's a dynamic file, it should be placed in /var
Think we should put a simbolic link in /etc to make the dynamic configuration easily accessible... 

About the stream part i think we should leave it to the users...

Anyway i was thinking of set the nginx.conf file entirely dynamic (Have a template and recreate it every time in the dynamic folder, this way it would be easier to add feature in the future, since the nginx.conf template should not be changed and should be provided by the package)
I hope i explained what i mean clearly.

**Comment by peter-stadler** [comment]
I see: So I will change the init, to create symbolic links on start and remove them on stop (restart and reload similarly).

No stream part is good for me :-)

The point is putting the whole nginx.conf to UCI is quite some work (to unify the different styles of configuration) and I would rather do it step by step. There are quite some decisions, how to refer between different parts and what to do with missing parts ...

Thank you very much, I think I understand you clearly.

**Comment by peter-stadler** [comment]
Invitation for discussion also to @peterwillcn @lucize @zxlhhyccc as you could give me already some valuable feedback: What do you think?
I know I could only adopt very little to your proposals, but I think I could make it at least more flexible.

**Comment by peter-stadler** [comment]
Ah, I re-read your comment: you mean the symlink for the main nginx.conf, don't you?

**Comment by Ansuel** [comment]
Yes like we do with samba

provide a nginx.conf.template and use it to generate the final dynamic nginx.conf (in the /var dir)

**Comment by peter-stadler** [comment]
I think I could add a `section main http` to the UCI config that contains the current options of `/etc/nginx/nginx.conf`.
**Edit:** I will look at how it is done for samba ...

**Comment by neheb** [comment]
Just saw that this includes several bash scripts. Please run through shellcheck and switch them to use /bin/sh.

**Comment by peter-stadler** [comment]
Ah, thank you. I oversaw that since the testscripts are meant to be run on the host ... I will change it.

**Comment by peter-stadler** [comment]
I changed it to setup the whole conf from a [template](https://github.com/openwrt/packages/blob/f37abf9f91159466efdc9637ecb02b52600cdac3/net/nginx-util/files/nginx.conf.template) by inserting the uci servers when invoking `nginx-util init_lan -uci`. Is this going to the right direction?

**Comment by peter-stadler** [comment]
For Nginx we could symlink `/etc/nginx/nginx.conf -> /var/lib/nginx/uci.conf` and init by:
```
	[ -d /var/log/nginx ] || mkdir -p /var/log/nginx
	[ -d /var/lib/nginx ] || mkdir -p /var/lib/nginx

	if [ "$(readlink "${NGINX_CONF}")" = "/var/lib/nginx/uci.conf" ]
	then ${NGINX_UTIL} init_lan -uci
	else ${NGINX_UTIL} init_lan
	fi

	nginx -t -c "${NGINX_CONF}" -g "daemon off; worker_processes 0;" || {
		echo "TODO"
		exit 1
	}
```

**Comment by Ansuel** [comment]
Yes i like the template way
I think create a symlink could cause some problem... Would be better to add a config in the general uci config to autogenerate a nginx conf or use a custom config

Then in init.d check for this option and load the appropriate config accordingly.

**Comment by peter-stadler** [comment]
Wouldn't it good to recognize if there is a user modified `nginx.conf` automatically, the symlink idea I got from samba. If it is only about the symlink, we could start nginx with different `-c` values; something like:
```
if [ -e "${NGINX_CONF}" ]
then ${NGINX_UTIL} init_lan && nginx -c "${NGINX_CONF}" ...
else ${NGINX_UTIL} init_lan -uci && nginx -c "${UCI_CONF}" ...
fi
```
What do you think?

Before I saw your comment I just added a prefix `uci_` to the special options `manage_ssl` and `listen_locally`. Only the options without this prefix are copied to the server part (assuming they are standard Nginx options). If we need to add more options in the future, we can prepend `uci_` to them and they will be ignored by old versions of the `nginx-util` ...

**Comment by peter-stadler** [comment]
Or how do you think we should do with Nginx?
I thought about adopting the init.d in a next step while removing the files `nginx.conf` and `_lan.conf` (and `_redirect2ssl.conf` for the SSL version) from the package. So, these files will be removed on the upgrade if the user has not modified them. I thought about using that as signal what the user wants, but I am also happy of using a UCI config option for it (What should be the default value then?)

**Comment by Ansuel** [comment]
> Or how do you think we should do with Nginx?
> I thought about adopting the init.d in a next step while removing the files `nginx.conf` and `_lan.conf` (and `_redirect2ssl.conf` for the SSL version) from the package. So, these files will be removed on the upgrade if the user has not modified them. I thought about using that as signal what the user wants, but I am also happy of using a UCI config option for it (What should be the default value then?) ...

well the default would be to use the uci configuration and the template to generate the final config...

**Comment by peter-stadler** [comment]
Some users will not be happy now if they are forced to opt-out again (this time from the UCI config). But, in the long run maybe it is better if a user has to be explicit, when using a custom `nginx.conf` ...

**Comment by Ansuel** [comment]
implementation of nginx on openwrt is special, this should not cause so much trouble... As it will take only one option to restore the original behavior

**Comment by peter-stadler** [comment]
Done. It is enabled and can be disabled by `uci set nginx.global.uci_enable=false` ...

**Comment by peter-stadler** [comment]
Thank you :-)

**Comment by peter-stadler** [comment]
Hmm, one problem: If we invoke `nginx -c /var/lib/nginx/uci.conf ...` directly, nginx uses the prefix `/var/lib/nginx/` as directory for relative includes (e.g. for `mime.types`).

Should we link all includables to the temporary folder `/var/lib/nginx/` when starting from init.d?

Or should we make (at install time a dead) symlink like `/etc/nginx/uci.conf -> /var/lib/nginx/uci.conf`? Then we create `/var/lib/nginx/uci.conf` dynamically, but start Nginx it via the symlink: `nginx -c /etc/nginx/uci.conf ...`

**Comment by peter-stadler** [comment]
Done the latter: Creating `/var/lib/nginx/uci.conf` dynamically, here.
The `nginx` package would be supposed to make (at install time a dead) symlink 
`/etc/nginx/uci.conf -> /var/lib/nginx/uci.conf`
and use that symlink in init.d ...

**Comment by peter-stadler** [comment]
tested and followed #11611

**Comment by peter-stadler** [comment]
(copied the description to the commit message)

**Comment by peter-stadler** [comment]
(rebased and) added documentation there is also the [draft for Nginx](11726) ...

**Comment by peter-stadler** [comment]
(corrected/improved documentation and rebased)

**Comment by peter-stadler** [comment]
@Ansuel what do you think?

**Comment by Ansuel** [comment]
@peter-stadler i need to test this. 
Do you test this in various situation? (for example, works with empty system, works with a system running the legacy version)
A very old commit that I tested broke my cron rule (as it was appended to the end of the line instead of adding it to a new line)

**Comment by peter-stadler** [comment]
Sure, no hurry :-) The question was more about the design.

I do manual tests and run the scripts src/test*.sh (and with clang-tidy beforehand). When changing memory management I test with valgrind, too.
Usually I do it for two different circumstances in qemu:
* Start a clean instance, install the new version.
* Start a clean instance, install the old version, upgrade to the new.

But, there are sure some cases I am missing. The cron job is one of them, I changed it now.

Thank you very much and if you have more suggestions (what to test, which tools to use, ...) just tell me :-)

**Comment by peter-stadler** [comment]
(made a clarification to the documentation and changed the test to save and recover the configs on the system: although it works, I do **not** recommend running the tests on a working system)

**Comment by peter-stadler** [comment]
Fixed a bug for a smoother transition: Remove also the certificate for `_lan` not only the directives for `_lan` when calling `nginx-util del_ssl` (without name). The `prerm` of Nginx expects it to do so (and this is the only situation, where it should be called; in the next OpenWrt release support only `nginx-util del_ssl $name` …)

**Comment by peter-stadler** [comment]
Changed `nginx_util [add_ssl|del_ssl] FQDN` for switching the port for the listen directives between `80` and `443 ssl` for server parts in `/etc/nginx/conf.d/FQDN.conf`. Now, these commands behaves similar as for the UCI configuration …

**Comment by peter-stadler** [comment]
Thinking about #11847 I am wondering: For restricting the access to the default_server locally, would it be better to listen on all IPs (instead of the loopback and lan IPs as now) and do something like the following?
```
allow ::1;
allow fc00::/7;
allow fec0::/10;
allow fe80::/10;
allow 127.0.0.0/8;
allow 10.0.0.0/8;
allow 172.16.0.0/12;
allow 192.168.0.0/16;
allow 169.254.0.0/16;
deny all;
```

**Comment by peter-stadler** [comment]
I am getting more and more convinced that it would be better to use allow/deny instead of listen:
`+` There is no need for `uci_listen_locally` or including `lan_ssl.listen`; other servers can just use `listen 80;` or `listen 443 ssl;`
`+` We do not reload Nginx on changes of the `lan` and `loopback` interfaces.
`+` Link local addresses would work if the browser supports them. (Opening an SSL tunnel works in all cases.)
`-` Other servers cannot use `listen 80 default_server;` resp. `listen 443 ssl default_server;` The `_lan` server is the only one then. There could be different `default_servers` for different listen IP:port values (as it would have been the case if listening locally).

The latter was the reason, I implemented it using `listen`. So one could add another `default_server` without changing the `_la server. But, now I think, this is a very rare use case (with SSL you would typically setup the `server_name` and Nginx is choosing it by the host name); and in that case, one would probably change the `_lan` server anyway …

**Comment by EricLuehrsen** [comment]
I think this does already, but to be sure. It is best to all config, cache, or whatever end up in `/var/lib/[server]`. This includes a stupid copy of `/etc/somthing.conf` if need be. A `procd` jail or an in process `chroot` is much simpler to configure.

**Comment by peter-stadler** [comment]
Currently only the dynamic files are created at `/var/lib/nginx`. The whole Nginx configuration is quite flexible (most of it would be in `/etc/nginx/`, but it could also include other files, e.g. from `/etc/ssl/`). If we want to recursively copy `/etc/nginx/` to `/var/lib/nginx`, I think it would be the best letting `/etc/init.d/nginx` do that.

How would a `procd` jail work? Would especially `luci-nginx` (with ubus) still work? (We need `user root;` in the Nginx configuration for that if I see it right.)

**Comment by Ansuel** [comment]
procd jail needs nginx to run in a separated user?

I need to search if ubus can run with non root users.

**Comment by Ansuel** [comment]
> I am getting more and more convinced that it would be better to use `allow/deny` instead of `listen`:
> `+` There is no need for `uci_listen_locally` or including `lan_ssl.listen`; other servers can just use `listen 80;` or `listen 443 ssl;`
> `+` We do not reload Nginx on changes of the `lan` and `loopback` interfaces.
> `+` Link local addresses would work if the browser supports them. (Opening an SSL tunnel works in all cases.)
> `-` Other servers cannot use `listen 80 default_server;` resp. `listen 443 ssl default_server;` The `_lan` server is the only one then. There could be different `default_servers` for different listen IP:port values (as it would have been the case if listening locally).
> 
> The latter was the reason, I implemented it using `listen`. So one could add another `default_server` without changing the `_lan` server. But, now I think, this is a very rare use case (with SSL you would typically setup the `server_name` and Nginx is choosing it by the host name); and in that case, one would probably change the `_lan` server anyway …

So with allow/deny we can have only one server that can listen on 80 or 443 right ?

**Comment by peter-stadler** [comment]
> So with allow/deny we can have only one server that can listen on 80 or 443 right ?

No, there can only be one `default_server`. Other servers can listen on 80 resp. 443, too; they  have to set the `server_name` to the domain to be useful, though.

**Comment by peter-stadler** [comment]
For making the point clear, here are two minimal examples for port 80 (with ssl it would be similar)

----
Example for current version:
```
server { # preset_lan
   listen 192.168.1.1:80 default_server;
}
server { # custom1
    listen 80 default_server;
}
server { # custom2
    listen 192.168.1.1:80;
    listen 80;
    server_name example.com;
}
```
* Any request with the host name `example.com` is processed by `custom2`.

For other requests it depends: 
* If the request is got **on** 192.168.1.1, then it is processed by `preset_lan`, 
* else by `custom1`. 

Any server has to listen explicitly on the LAN address to be reachable locally (on port 80), a `listen *:80;` is not enough: the `custom2` part needs the `listen 192.168.1.1:80;` line for that.

----
Example for other version:
```
server { # preset_lan
    listen 80 default_server;
    allow 192.168.0.0/16;
    deny all;
}
# server { # custom 1
#     listen 80 default_server;
# }
server { # custom2
   listen 80;
   server_name example.com;
}
```
* Any request with the host name `example.com` is processed by `custom2`.
* All other request are passed to `preset_lan`: It allows only those request which come **from** a LAN address (starting with 192.168).

It is [invalid](https://nginx.org/en/docs/http/request_processing.html) to have more than one `default_server` for any IP:PORT combination: uncommenting the `custom1` part would lead to an error.

**Comment by peter-stadler** [comment]
For the procd jail I am not sure as I do not know how it would work.

I think a jail would restrict somehow, what the process is allowed to do; this could interfere with the needs of LuCI. But, I would expect that it is something else than setting the user …

If you can point me how to set up a jail, I would test it.

The [wiki](https://openwrt.org/docs/techref/ubus#acls) talks about ACLs for ubus, but I did not look into that. And, I only understand very little about how LuCI on Nginx is working …

**Comment by peter-stadler** [comment]
rebased and fixed #12273

**Comment by neheb** [comment]
On an unrelated note, I looked through the code and found some low hanging fruit:

```
--- a/net/nginx-util/src/nginx-util.cpp
+++ b/net/nginx-util/src/nginx-util.cpp
@@ -1,4 +1,5 @@
 #include <iostream>
+#include <numeric>
 
 #include "nginx-util.hpp"
 
@@ -170,11 +171,9 @@ auto main(int argc, char * argv[]) -> int
 #endif
             std::cerr<<"pthread and libstdcpp)."<<std::endl;
 
-            auto usage = std::string{"usage: "} + *argv + " [";
-            for (auto cmd : cmds) {
-                usage += std::string{cmd[0]};
-                usage += std::string{cmd[1]} + "|";
-            }
+            auto usage = std::accumulate(cmds.begin(), cmds.end(), std::string{"usage: "} + *argv + " [",
+                [](const auto &use, const auto &cmd){ return use + std::string{cmd[0]} +
+                                                             use + std::string{cmd[1]} + "|"; });
             usage[usage.size()-1] = ']';
             std::cerr<<usage<<std::endl;
 
--- a/net/nginx-util/src/px5g.cpp
+++ b/net/nginx-util/src/px5g.cpp
@@ -1,6 +1,7 @@
 #include "px5g-openssl.hpp"
 #include <array>
 #include <iostream>
+#include <numeric>
 #include <string>
 #include <string_view>
 #include <unistd.h>
@@ -422,10 +423,8 @@ auto main(int argc, const char ** argv) -> int
 
     catch (const std::exception & e)  {
 
-        auto usage = std::string{"usage: \n"}  ;
-        for (auto cmd : cmds) {
-            usage += std::string{4, ' '} + *argv +" "+ cmd[0] + cmd[1] +"\n";
-        }
+        auto usage = std::accumulate(cmds.begin(), cmds.end(), std::string{"usage: \n"},
+            [=](const auto &use, const auto &cmd){ return use + std::string{4, ' '} + *argv +" "+ cmd[0] + cmd[1] +"\n"; });
 
         std::cerr<<usage<<std::flush;
```

Not sure if useful. I do know accumulate generates better code. At least with -O2.

**Comment by peter-stadler** [comment]
Thank you, I integrated these changes (I do not see a big difference) :-)

**Comment by peter-stadler** [comment]
Rebased and added another commit restricting the access to LAN with allow/deny directives (instead of using listen locally as before).
I made a new commit for tracking the differences and would squash the two commits later …

**Comment by jkcapoeira** [comment]
I have sslh running as multiplexer on port 443 and forward tls to nginx which is running as a reverse proxy on port 9443. So i can log for each virtual machine the real ip address. (https://forum.openwrt.org/t/nginx-and-openvpn-on-port-443/25004/5) but to get that working again i think i need to replace the /ect/init.d/nginx with an old version where there is no reference to nginx-util and all the LAN_stuff correct?

**Comment by peter-stadler** [comment]
Squashed the two commits, rebased and tested.

**Comment by Ansuel** [comment]
@peter-stadler what test did you do? 
From what i can read if a config is already present the generation is skipped right?

**Comment by peter-stadler** [comment]
I did run the `test-nginx-util.sh` and opened the web interface with `luci-nginx` after freshly installing and upgrading both this tool and a modified [nginx (PR: use UCI config provided by nginx-util)](https://github.com/openwrt/packages/pull/11726). I did run also the following on a **test system (this alters the config)**:
```
test_nginx() { # working after reboot?!
    [ -e /var/run/nginx.pid ] || return 5 #running after install.
    uci revert nginx && service nginx reload && rm -f /etc/nginx/nginx.conf
    ls /etc/nginx/* /var/lib/nginx/ && cat /etc/config/nginx
    cat /etc/nginx/uci.conf >/nginx.conf || return 4 #prepare.
    uci set nginx.global.uci_enable=0 || return 3 #disable.
    service nginx reload && return 2 #should complain about nginx.conf
    mv /nginx.conf /etc/nginx/ && service nginx reload || return 1 #should work
    logread | grep nginx #recapitulate
}
test_nginx; echo $?
```

The upgrade is a two step process:
1. Upgrade nginx-util: Makes UCI config available; the new configuration will be generated at `/var/lib/nginx/uci.conf` from the template, but nginx still uses the old files. It skips also the generation of the `_lan` (and the `_redirect2ssl`) server part since these files are present as `/etc/nginx/conf.d/*.conf` …
2. Upgrade Nginx: Changes the init.d script for using the new (generated) configuration file by default. And also removes `_lan.conf` (as well as `_redirect2ssl.conf`) if not modified; then these server parts will be generated from the UCI config.

Simulating the upgrade of Nginx can be done by `rm /etc/nginx/conf.d/_lan.conf` and replacing `/etc/init.d/nginx` with:
```
#!/bin/sh /etc/rc.common
# Copyright (C) 2015 OpenWrt.org

START=80

USE_PROCD=1

G_OPTS="daemon off;"

NGINX_UTIL="/usr/bin/nginx-util"

eval $("${NGINX_UTIL}" get_env)

CONF=""


nginx_init() {
	[ -z "${CONF}" ] || return # already called.

	[ -d /var/log/nginx ] || mkdir -p /var/log/nginx
	[ -d /var/lib/nginx ] || mkdir -p /var/lib/nginx

	rm -f "$(readlink "${UCI_CONF}")"
	${NGINX_UTIL} init_lan

	if [ -e "${UCI_CONF}" ]
	then CONF="${UCI_CONF}"
	else CONF="${NGINX_CONF}"
	fi

	local message
	message="$(/usr/sbin/nginx -t -c "${CONF}" -g "${G_OPTS}" 2>&1)" ||
	{
		echo -e "${message}" | logger -t "nginx_init" -p "daemon.err"
		logger -s -t "nginx_init" -p "daemon.err" "NOT using conf file!"
		echo "show config to be used by: nginx -T -c '${CONF}'" >&2
		exit 1
	}

	logger -t "nginx_init" -p "daemon.info" "using ${CONF} (the test is ok)"
}


start_service() {
	nginx_init

	procd_open_instance
	procd_set_param command /usr/sbin/nginx -c "${CONF}" -g "${G_OPTS}"
	procd_set_param stdout 1
	procd_set_param stderr 1
	procd_set_param file "${CONF}" "${CONF_DIR}*.crt" "${CONF_DIR}*.key" \
		"${CONF_DIR}*.conf" "${CONF_DIR}*.locations"
	procd_set_param respawn
	procd_close_instance
}


reload_service() {
	nginx_init

	if [ "$(cat "/proc/$(cat "/var/run/nginx.pid")/cmdline")" = \
	     "nginx: master process /usr/sbin/nginx -c ${CONF} -g ${G_OPTS}" ]
	then procd_send_signal nginx
	else restart
	fi
}


relog() {
	[ -d /var/log/nginx ] || mkdir -p /var/log/nginx
	procd_send_signal nginx '*' USR1
}


EXTRA_COMMANDS="relog"
EXTRA_HELP="	relog	Reopen log files (without reloading)"
```

**Comment by peter-stadler** [comment]
To make it clear: Yes, a server part of the UCI config is skipped iff there is a `.conf` file with the same name in /etc/nginx/conf.d/, e.g., a `config server 'example_com'` section will be skipped when there is a `example_com.conf` file (but not when the file is named `example.com.conf`).

**Comment by peter-stadler** [comment]
(back to draft: wait until the the nossl version of nginx is removed; this will change the `README.sh` here)

**Comment by majkrzak** [comment]
Were you consider usecases of "reverse proxy" and "acme" certificates?

**Comment by peter-stadler** [comment]
I did consider it. The reverse proxy use case should be creating an appropriate `/etc/nginx/conf.d/$name.locations` file, it will be included by default in the `_lan` server (that now is in the `/etc/nginx/conf.d/_lan.conf` and after this PR will be in the UCI config `/etc/config/nginx`). Other server parts can include the file, too. For example the packages `luci-nginx` and `etesync-server` are creating corresponding `.locations` files.

I thought also about `acme` and saw the problem that it is trying to edit the main file (`/etc/nginx/nginx.conf`). I think it would be best to change it after this is completed (to edit the corresponding UCI server part or the `/etc/nginx/conf.d/${FQDN}.conf` file). I thought about adding support for it to `nginx-util add_ssl …`, too …

**Comment by peter-stadler** [comment]
I did consider it. The reverse proxy use case should be creating an appropriate `/etc/nginx/conf.d/$name.locations` file, it will be included by default in the `_lan` server (that now is in the `/etc/nginx/conf.d/_lan.conf` and after this PR will be in the UCI config `/etc/config/nginx`). Other server parts can include the file, too. For example the packages `luci-nginx` and `etesync-server` are creating corresponding `.locations` files.

I thought also about `acme` and saw the problem that it is trying to edit the main file (`/etc/nginx/nginx.conf`). I think it would be best to change it after this is completed (to edit the corresponding UCI server part or the `/etc/nginx/conf.d/${FQDN}.conf` file). I thought about adding support for it to `nginx-util add_ssl …`, too …

**Comment by peter-stadler** [comment]
Will make another PR after #12929
